modules/derivation/shared/src/main/scala/io/circe/derivation/DerivationMacros.scala

```diff
@@ -678,39 +692,40 @@ class DerivationMacros(val c: blackbox.Context) extends ScalaVersionCompat {
     }
 
   private[this] def materializeCodecTraitImpl[T: c.WeakTypeTag](
-    transformMemberNames: Option[c.Expr[String => String]],
+    transformConstructorNames: Option[c.Expr[String => String]],
     useDefaults: c.Expr[Boolean],
     discriminator: c.Expr[Option[String]],
     subclasses: Set[Symbol]
   ): c.Expr[Codec.AsObject[T]] = {
     val tpe = weakTypeOf[T]
 
+    def transformName(name: String): Tree = transformConstructorNames.fold[Tree](q"$name")(f => q"$f($name)")
+
     val objectWrapperCases: List[Tree] = subclasses.map { s =>
-      val value =
-        Literal(Constant(nameOf(s)))
+      val value = transformName(nameOf(s))
 
-      cq"""$value => c.get($value)(_root_.io.circe.Decoder[${s.asType}]).asInstanceOf[_root_.io.circe.Decoder.Result[$tpe]]"""
+      cq"""nme if nme == $value => c.get(nme)(_root_.io.circe.Decoder[${s.asType}]).asInstanceOf[_root_.io.circe.Decoder.Result[$tpe]]"""
```

Comment:
Out of curiosity, was this change necessary?

Reply:
It was (as far as I could tell). Matching directly on `$value` caused compilation errors about function 1 not having an `unapply` method when `transformConstructorNames` wasn't `None`. The generated code looked like:
```scala
case (((x: String) => someFunc(x))("AdtBar")) => ???
case (((x: String) => someFunc(x))("AdtFoo")) => ???
```
Now they look like:
```scala
case nme if nme == (((x: String) => someFunc(x))("AdtBar")) => ???
case nme if nme == (((x: String) => someFunc(x))("AdtFoo")) => ???
```

Related question I was wondering about: is there any way to "lift" the transformation function into the macro to make it directly useable, so we can pre-compute the transformed strings?